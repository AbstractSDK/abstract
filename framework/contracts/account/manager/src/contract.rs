use crate::{
    commands::*,
    error::ManagerError,
    queries,
    queries::{handle_account_info_query, handle_config_query, handle_module_info_query},
    versioning,
};
use abstract_sdk::core::{
    manager::{
        state::{AccountInfo, Config, ACCOUNT_FACTORY, CONFIG, INFO, SUSPENSION_STATUS},
        CallbackMsg, ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg,
    },
    objects::module_version::assert_contract_upgrade,
    objects::validation::{validate_description, validate_link, validate_name},
    proxy::state::ACCOUNT_ID,
    MANAGER,
};
use cosmwasm_std::{
    ensure_eq, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdError, StdResult,
};
use cw2::set_contract_version;
use semver::Version;

pub type ManagerResult<R = Response> = Result<R, ManagerError>;

pub const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");

#[cfg_attr(feature = "export", cosmwasm_std::entry_point)]
pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> ManagerResult {
    let version: Version = CONTRACT_VERSION.parse().unwrap();

    assert_contract_upgrade(deps.storage, MANAGER, version)?;
    set_contract_version(deps.storage, MANAGER, CONTRACT_VERSION)?;
    Ok(ManagerResponse::action("migrate"))
}

#[cfg_attr(feature = "export", cosmwasm_std::entry_point)]
pub fn instantiate(
    deps: DepsMut,
    env: Env,
    info: MessageInfo,
    msg: InstantiateMsg,
) -> ManagerResult {
    set_contract_version(deps.storage, MANAGER, CONTRACT_VERSION)?;

    ACCOUNT_ID.save(deps.storage, &msg.account_id)?;
    CONFIG.save(
        deps.storage,
        &Config {
            version_control_address: deps.api.addr_validate(&msg.version_control_address)?,
            module_factory_address: deps.api.addr_validate(&msg.module_factory_address)?,
        },
    )?;

    // Verify info
    validate_description(&msg.description)?;
    validate_link(&msg.link)?;
    validate_name(&msg.name)?;

    let governance_details = msg.owner.verify(deps.api)?;
    let owner = governance_details.owner_address();

    let account_info = AccountInfo {
        name: msg.name,
        governance_details,
        chain_id: env.block.chain_id,
        description: msg.description,
        link: msg.link,
    };

    INFO.save(deps.storage, &account_info)?;
    MIGRATE_CONTEXT.save(deps.storage, &vec![])?;

    // Set owner
    cw_ownable::initialize_owner(deps.storage, deps.api, Some(owner.as_str()))?;
    SUSPENSION_STATUS.save(deps.storage, &false)?;
    ACCOUNT_FACTORY.set(deps, Some(info.sender))?;
    Ok(ManagerResponse::new(
        "instantiate",
        vec![
            ("account_id", msg.account_id.to_string()),
            ("owner", owner.to_string()),
        ],
    ))
}

#[cfg_attr(feature = "export", cosmwasm_std::entry_point)]
pub fn execute(deps: DepsMut, env: Env, info: MessageInfo, msg: ExecuteMsg) -> ManagerResult {
    match msg {
        ExecuteMsg::UpdateStatus {
            is_suspended: suspension_status,
        } => update_account_status(deps, info, suspension_status),
        msg => {
            // Block actions if user is not subscribed
            let is_suspended = SUSPENSION_STATUS.load(deps.storage)?;
            if is_suspended {
                return Err(ManagerError::AccountSuspended {});
            }

            match msg {
                ExecuteMsg::UpdateInternalConfig(config) => {
                    update_internal_config(deps, info, config)
                }
                ExecuteMsg::SetOwner { owner } => set_owner(deps, env, info, owner),

                ExecuteMsg::InstallModule { module, init_msg } => {
                    install_module(deps, info, env, module, init_msg)
                }
                ExecuteMsg::UninstallModule { module_id } => {
                    uninstall_module(deps, info, module_id)
                }
                ExecuteMsg::RegisterModule {
                    module,
                    module_addr,
                } => register_module(deps, info, env, module, module_addr),
                ExecuteMsg::ExecOnModule {
                    module_id,
                    exec_msg,
                } => exec_on_module(deps, info, module_id, exec_msg),
                ExecuteMsg::CreateSubAccount {
                    name,
                    description,
                    link,
                } => create_subaccount(deps, env, info, name, description, link),
                ExecuteMsg::Upgrade { modules } => upgrade_modules(deps, env, info, modules),
                ExecuteMsg::UpdateInfo {
                    name,
                    description,
                    link,
                } => update_info(deps, info, name, description, link),
                ExecuteMsg::UpdateSettings {
                    ibc_enabled: new_status,
                } => {
                    let mut response: Response = ManagerResponse::action("update_settings");

                    if let Some(ibc_enabled) = new_status {
                        response = update_ibc_status(deps, info, ibc_enabled, response)?;
                    } else {
                        return Err(ManagerError::NoUpdates {});
                    }

                    Ok(response)
                }
                ExecuteMsg::Callback(CallbackMsg {}) => handle_callback(deps, env, info),
                ExecuteMsg::UpdateOwnership(action) => match action {
                    // Disallow the user from using the TransferOwnership action
                    cw_ownable::Action::TransferOwnership { .. } => {
                        Err(ManagerError::MustUseSetOwner {})
                    }
                    _ => {
                        abstract_sdk::execute_update_ownership!(
                            ManagerResponse,
                            deps,
                            env,
                            info,
                            action
                        )
                    }
                },
                _ => panic!(),
            }
        }
    }
}

#[cfg_attr(feature = "export", cosmwasm_std::entry_point)]
pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
    match msg {
        QueryMsg::ModuleVersions { ids } => queries::handle_contract_versions_query(deps, env, ids),
        QueryMsg::ModuleAddresses { ids } => queries::handle_module_address_query(deps, env, ids),
        QueryMsg::ModuleInfos { start_after, limit } => {
            handle_module_info_query(deps, start_after, limit)
        }
        QueryMsg::Info {} => handle_account_info_query(deps),
        QueryMsg::Config {} => handle_config_query(deps),
        QueryMsg::Ownership {} => abstract_sdk::query_ownership!(deps),
    }
}

pub fn handle_callback(mut deps: DepsMut, env: Env, info: MessageInfo) -> ManagerResult {
    ensure_eq!(
        info.sender,
        env.contract.address,
        StdError::generic_err("Callback must be called by contract")
    );
    let migrated_modules = MIGRATE_CONTEXT.load(deps.storage)?;

    for (migrated_module_id, old_deps) in migrated_modules {
        versioning::maybe_remove_old_deps(deps.branch(), &migrated_module_id, &old_deps)?;
        let new_deps =
            versioning::maybe_add_new_deps(deps.branch(), &migrated_module_id, &old_deps)?;
        versioning::assert_dependency_requirements(deps.as_ref(), &new_deps, &migrated_module_id)?;
    }

    MIGRATE_CONTEXT.save(deps.storage, &vec![])?;
    Ok(Response::new())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contract;
    use cosmwasm_std::testing::*;
    use speculoos::prelude::*;

    use crate::test_common::mock_init;

    mod migrate {
        use super::*;
        use abstract_core::AbstractError;
        use cw2::get_contract_version;

        #[test]
        fn disallow_same_version() -> ManagerResult<()> {
            let mut deps = mock_dependencies();
            mock_init(deps.as_mut())?;

            let version: Version = CONTRACT_VERSION.parse().unwrap();

            let res = contract::migrate(deps.as_mut(), mock_env(), MigrateMsg {});

            assert_that!(res)
                .is_err()
                .is_equal_to(ManagerError::Abstract(
                    AbstractError::CannotDowngradeContract {
                        contract: MANAGER.to_string(),
                        from: version.clone(),
                        to: version,
                    },
                ));

            Ok(())
        }

        #[test]
        fn disallow_downgrade() -> ManagerResult<()> {
            let mut deps = mock_dependencies();
            mock_init(deps.as_mut())?;

            let big_version = "999.999.999";
            set_contract_version(deps.as_mut().storage, MANAGER, big_version)?;

            let version: Version = CONTRACT_VERSION.parse().unwrap();

            let res = contract::migrate(deps.as_mut(), mock_env(), MigrateMsg {});

            assert_that!(res)
                .is_err()
                .is_equal_to(ManagerError::Abstract(
                    AbstractError::CannotDowngradeContract {
                        contract: MANAGER.to_string(),
                        from: big_version.parse().unwrap(),
                        to: version,
                    },
                ));

            Ok(())
        }

        #[test]
        fn disallow_name_change() -> ManagerResult<()> {
            let mut deps = mock_dependencies();
            mock_init(deps.as_mut())?;

            let old_version = "0.0.0";
            let old_name = "old:contract";
            set_contract_version(deps.as_mut().storage, old_name, old_version)?;

            let res = contract::migrate(deps.as_mut(), mock_env(), MigrateMsg {});

            assert_that!(res)
                .is_err()
                .is_equal_to(ManagerError::Abstract(
                    AbstractError::ContractNameMismatch {
                        from: old_name.parse().unwrap(),
                        to: MANAGER.parse().unwrap(),
                    },
                ));

            Ok(())
        }

        #[test]
        fn works() -> ManagerResult<()> {
            let mut deps = mock_dependencies();
            mock_init(deps.as_mut())?;

            let small_version = "0.0.0";
            set_contract_version(deps.as_mut().storage, MANAGER, small_version)?;

            let version: Version = CONTRACT_VERSION.parse().unwrap();

            let res = contract::migrate(deps.as_mut(), mock_env(), MigrateMsg {})?;
            assert_that!(res.messages).has_length(0);

            assert_that!(get_contract_version(&deps.storage)?.version)
                .is_equal_to(version.to_string());
            Ok(())
        }
    }
}
