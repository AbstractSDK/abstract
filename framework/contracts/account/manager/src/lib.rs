mod commands;
pub mod contract;
pub mod error;
mod queries;
mod validation;
mod versioning;

#[cfg(test)]
mod test_common {
    use abstract_core::{
        manager,
        objects::{account::AccountTrace, gov_type::GovernanceDetails, AccountId},
    };
    use abstract_testing::prelude::*;
    use cosmwasm_std::{testing::*, DepsMut};

    use crate::contract::ManagerResult;

    /// Initialize the manager with the test owner as the owner
    pub(crate) fn mock_init(mut deps: DepsMut) -> ManagerResult {
        let info = mock_info(TEST_ACCOUNT_FACTORY, &[]);

        crate::contract::instantiate(
            deps.branch(),
            mock_env(),
            info,
            manager::InstantiateMsg {
                account_id: AccountId::new(1, AccountTrace::Local).unwrap(),
                owner: GovernanceDetails::Monarchy {
                    monarch: OWNER.to_string(),
                },
                version_control_address: TEST_VERSION_CONTROL.to_string(),
                module_factory_address: TEST_MODULE_FACTORY.to_string(),
                proxy_addr: TEST_PROXY.to_string(),
                name: "test".to_string(),
                description: None,
                link: None,
                install_modules: vec![],
            },
        )
    }
}
