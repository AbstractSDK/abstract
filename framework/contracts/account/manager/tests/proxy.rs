mod common;
use abstract_adapter::mock::MockExecMsg;
use abstract_core::adapter::AdapterRequestMsg;
use abstract_core::manager::ModuleVersionsResponse;
use abstract_core::module_factory::ModuleInstallConfig;
use abstract_core::objects::fee::FixedFee;
use abstract_core::objects::module::{ModuleInfo, ModuleVersion, Monetization};
use abstract_core::objects::module_reference::ModuleReference;
use abstract_core::objects::namespace::Namespace;
use abstract_core::version_control::UpdateModule;
use abstract_core::{manager::ManagerModuleInfo, PROXY};
use abstract_interface::*;
use abstract_manager::contract::CONTRACT_VERSION;
use abstract_testing::prelude::{TEST_ACCOUNT_ID, TEST_MODULE_ID};
use common::{
    create_default_account, init_mock_adapter, install_adapter, mock_modules, AResult, TEST_COIN,
};
use cosmwasm_std::{coin, to_binary, wasm_execute, Addr, Coin, CosmosMsg};
use cw2::ContractVersion;
use cw_orch::deploy::Deploy;
use cw_orch::prelude::*;
use module_factory::error::ModuleFactoryError;
use speculoos::prelude::*;

#[test]
fn instantiate() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain, sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;

    let modules = account.manager.module_infos(None, None)?.module_infos;

    // assert proxy module
    assert_that!(&modules).has_length(1);
    assert_that(&modules[0]).is_equal_to(&ManagerModuleInfo {
        address: account.proxy.address()?,
        id: PROXY.to_string(),
        version: cw2::ContractVersion {
            contract: PROXY.into(),
            version: CONTRACT_VERSION.into(),
        },
    });

    // assert manager config
    assert_that!(account.manager.config()?).is_equal_to(abstract_core::manager::ConfigResponse {
        version_control_address: deployment.version_control.address()?,
        module_factory_address: deployment.module_factory.address()?,
        account_id: TEST_ACCOUNT_ID.into(),
        is_suspended: false,
    });
    Ok(())
}

#[test]
fn exec_through_manager() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;

    // mint coins to proxy address
    chain.set_balance(
        &account.proxy.address()?,
        vec![Coin::new(100_000, TEST_COIN)],
    )?;

    // burn coins from proxy
    let proxy_balance = chain
        .app
        .borrow()
        .wrap()
        .query_all_balances(account.proxy.address()?)?;
    assert_that!(proxy_balance).is_equal_to(vec![Coin::new(100_000, TEST_COIN)]);

    let burn_amount: Vec<Coin> = vec![Coin::new(10_000, TEST_COIN)];

    account.manager.exec_on_module(
        cosmwasm_std::to_binary(&abstract_core::proxy::ExecuteMsg::ModuleAction {
            msgs: vec![CosmosMsg::Bank(cosmwasm_std::BankMsg::Burn {
                amount: burn_amount,
            })],
        })?,
        PROXY.to_string(),
        &[],
    )?;

    let proxy_balance = chain
        .app
        .borrow()
        .wrap()
        .query_all_balances(account.proxy.address()?)?;
    assert_that!(proxy_balance).is_equal_to(vec![Coin::new(100_000 - 10_000, TEST_COIN)]);

    Ok(())
}

#[test]
fn default_without_response_data() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    let _staking_adapter_one = init_mock_adapter(chain.clone(), &deployment, None)?;

    install_adapter(&account.manager, TEST_MODULE_ID)?;

    chain.set_balance(
        &account.proxy.address()?,
        vec![Coin::new(100_000, TEST_COIN)],
    )?;

    let resp = account.manager.execute_on_module(
        TEST_MODULE_ID,
        Into::<abstract_core::adapter::ExecuteMsg<MockExecMsg>>::into(MockExecMsg),
    )?;
    assert_that!(resp.data).is_none();

    Ok(())
}

#[test]
fn with_response_data() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = create_default_account(&deployment.account_factory)?;
    let staking_adapter = init_mock_adapter(chain.clone(), &deployment, None)?;

    install_adapter(&account.manager, TEST_MODULE_ID)?;

    staking_adapter
        .call_as(&account.manager.address()?)
        .execute(
            &abstract_core::adapter::ExecuteMsg::<MockExecMsg, Empty>::Base(
                abstract_core::adapter::BaseExecuteMsg::UpdateAuthorizedAddresses {
                    to_add: vec![account.proxy.addr_str()?],
                    to_remove: vec![],
                },
            ),
            None,
        )?;

    chain.set_balance(
        &account.proxy.address()?,
        vec![Coin::new(100_000, TEST_COIN)],
    )?;

    let adapter_addr = account
        .manager
        .module_info(TEST_MODULE_ID)?
        .expect("test module installed");
    // proxy should be final executor because of the reply
    let resp = account.manager.exec_on_module(
        cosmwasm_std::to_binary(&abstract_core::proxy::ExecuteMsg::ModuleActionWithData {
            // execute a message on the adapter, which sets some data in its response
            msg: wasm_execute(
                adapter_addr.address,
                &abstract_core::adapter::ExecuteMsg::<MockExecMsg, Empty>::Module(
                    AdapterRequestMsg {
                        proxy_address: Some(account.proxy.addr_str()?),
                        request: MockExecMsg,
                    },
                ),
                vec![],
            )?
            .into(),
        })?,
        PROXY.to_string(),
        &[],
    )?;

    let response_data_attr_present = resp.event_attr_value("wasm-abstract", "response_data")?;
    assert_that!(response_data_attr_present).is_equal_to("true".to_string());

    Ok(())
}

#[test]
fn install_standalone_modules() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = AbstractAccount::new(&deployment, Some(0));

    let standalone1_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_cw2::mock_execute,
        mock_modules::standalone_cw2::mock_instantiate,
        mock_modules::standalone_cw2::mock_query,
    ));
    let standalone1_id = chain.app.borrow_mut().store_code(standalone1_contract);

    let standalone2_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_no_cw2::mock_execute,
        mock_modules::standalone_no_cw2::mock_instantiate,
        mock_modules::standalone_no_cw2::mock_query,
    ));
    let standalone2_id = chain.app.borrow_mut().store_code(standalone2_contract);

    // install first standalone
    deployment.version_control.propose_modules(vec![(
        ModuleInfo {
            namespace: Namespace::new("abstract")?,
            name: "standalone1".to_owned(),
            version: ModuleVersion::Version(mock_modules::V1.to_owned()),
        },
        ModuleReference::Standalone(standalone1_id),
    )])?;

    account.install_module(
        "abstract:standalone1",
        &mock_modules::standalone_cw2::MockMsg,
        None,
    )?;

    // install second standalone
    deployment.version_control.propose_modules(vec![(
        ModuleInfo {
            namespace: Namespace::new("abstract")?,
            name: "standalone2".to_owned(),
            version: ModuleVersion::Version(mock_modules::V1.to_owned()),
        },
        ModuleReference::Standalone(standalone2_id),
    )])?;

    account.install_module(
        "abstract:standalone2",
        &mock_modules::standalone_no_cw2::MockMsg,
        None,
    )?;
    Ok(())
}

#[test]
fn install_standalone_versions_not_met() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = AbstractAccount::new(&deployment, Some(0));

    let standalone1_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_cw2::mock_execute,
        mock_modules::standalone_cw2::mock_instantiate,
        mock_modules::standalone_cw2::mock_query,
    ));
    let standalone1_id = chain.app.borrow_mut().store_code(standalone1_contract);

    // install first standalone
    deployment.version_control.propose_modules(vec![(
        ModuleInfo {
            namespace: Namespace::new("abstract")?,
            name: "standalone1".to_owned(),
            version: ModuleVersion::Version(mock_modules::V2.to_owned()),
        },
        ModuleReference::Standalone(standalone1_id),
    )])?;

    let err = account
        .install_module(
            "abstract:standalone1",
            &mock_modules::standalone_cw2::MockMsg,
            None,
        )
        .unwrap_err();

    if let AbstractInterfaceError::Orch(err) = err {
        let err: ModuleFactoryError = err.downcast()?;
        assert_eq!(
            err,
            ModuleFactoryError::Abstract(abstract_core::AbstractError::UnequalModuleData {
                cw2: mock_modules::V1.to_owned(),
                module: mock_modules::V2.to_owned(),
            })
        );
    } else {
        panic!("wrong error type")
    };

    Ok(())
}

#[test]
fn install_multiple_modules() -> AResult {
    let sender = Addr::unchecked(common::OWNER);
    let chain = Mock::new(&sender);
    chain.add_balance(&sender, vec![coin(86, "token1"), coin(500, "token2")])?;
    let deployment = Abstract::deploy_on(chain.clone(), sender.to_string())?;
    let account = AbstractAccount::new(&deployment, Some(0));

    let standalone1_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_cw2::mock_execute,
        mock_modules::standalone_cw2::mock_instantiate,
        mock_modules::standalone_cw2::mock_query,
    ));
    let standalone1_id = chain.app.borrow_mut().store_code(standalone1_contract);

    let standalone2_contract = Box::new(ContractWrapper::new(
        mock_modules::standalone_no_cw2::mock_execute,
        mock_modules::standalone_no_cw2::mock_instantiate,
        mock_modules::standalone_no_cw2::mock_query,
    ));
    let standalone2_id = chain.app.borrow_mut().store_code(standalone2_contract);

    // install both standalone
    deployment.version_control.propose_modules(vec![
        (
            ModuleInfo {
                namespace: Namespace::new("abstract")?,
                name: "standalone1".to_owned(),
                version: ModuleVersion::Version(mock_modules::V1.to_owned()),
            },
            ModuleReference::Standalone(standalone1_id),
        ),
        (
            ModuleInfo {
                namespace: Namespace::new("abstract")?,
                name: "standalone2".to_owned(),
                version: ModuleVersion::Version(mock_modules::V1.to_owned()),
            },
            ModuleReference::Standalone(standalone2_id),
        ),
    ])?;

    // add monetization on module1
    let monetization = Monetization::InstallFee(FixedFee::new(&coin(42, "token1")));
    deployment.version_control.update_module_configuration(
        "standalone1".to_owned(),
        Namespace::new("abstract").unwrap(),
        UpdateModule::Versioned {
            version: mock_modules::V1.to_owned(),
            metadata: None,
            monetization: Some(monetization),
            instantiation_funds: None,
        },
    )?;

    // add init funds on module2
    deployment.version_control.update_module_configuration(
        "standalone2".to_owned(),
        Namespace::new("abstract").unwrap(),
        UpdateModule::Versioned {
            version: mock_modules::V1.to_owned(),
            metadata: None,
            monetization: None,
            instantiation_funds: Some(vec![coin(42, "token1"), coin(500, "token2")]),
        },
    )?;

    // Don't allow to attach too much funds
    let err = account
        .install_modules(
            vec![
                ModuleInstallConfig::new(
                    ModuleInfo::from_id_latest("abstract:standalone1")?,
                    Some(to_binary(&mock_modules::standalone_cw2::MockMsg).unwrap()),
                ),
                ModuleInstallConfig::new(
                    ModuleInfo::from_id_latest("abstract:standalone2")?,
                    Some(to_binary(&mock_modules::standalone_no_cw2::MockMsg).unwrap()),
                ),
            ],
            Some(&[coin(86, "token1"), coin(500, "token2")]),
        )
        .unwrap_err();
    assert!(err.root().to_string().contains(&format!(
        "Expected {:?}, sent {:?}",
        vec![coin(84, "token1"), coin(500, "token2")],
        vec![coin(86, "token1"), coin(500, "token2")]
    )));

    // successful install
    account.install_modules_auto(vec![
        ModuleInstallConfig::new(
            ModuleInfo::from_id_latest("abstract:standalone1")?,
            Some(to_binary(&mock_modules::standalone_cw2::MockMsg).unwrap()),
        ),
        ModuleInstallConfig::new(
            ModuleInfo::from_id_latest("abstract:standalone2")?,
            Some(to_binary(&mock_modules::standalone_no_cw2::MockMsg).unwrap()),
        ),
    ])?;

    // Make sure all installed
    let account_module_versions = account.manager.module_versions(vec![
        String::from("abstract:standalone1"),
        // Querying no_cw2 fails
        // String::from("abstract:standalone2"),
    ])?;
    assert_eq!(
        account_module_versions,
        ModuleVersionsResponse {
            versions: vec![
                ContractVersion {
                    contract: String::from(mock_modules::standalone_cw2::MOCK_STANDALONE_ID),
                    version: String::from(mock_modules::V1)
                },
                // ContractVersion {
                //     contract: String::from("abstract:standalone2"),
                //     version: String::from(mock_modules::V1)
                // },
            ]
        }
    );

    let account_module_addresses = account.manager.module_addresses(vec![
        String::from("abstract:standalone1"),
        String::from("abstract:standalone2"),
    ])?;
    let (standalone_addr1, standalone_addr2) = match &account_module_addresses.modules[..] {
        [(_app1, addr1), (_app2, addr2)] => (addr1.clone(), addr2.clone()),
        _ => panic!("bad result from module_addresses"),
    };
    let s1_balance = chain.query_all_balances(&standalone_addr1)?;
    let s2_balance = chain.query_all_balances(&standalone_addr2)?;

    assert!(s1_balance.is_empty());
    assert_eq!(s2_balance, vec![coin(42, "token1"), coin(500, "token2")]);
    Ok(())
}
