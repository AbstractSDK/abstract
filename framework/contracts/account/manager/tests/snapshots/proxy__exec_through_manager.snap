---
source: contracts/account/manager/tests/proxy.rs
expression: all_storage
---
"abstract:account-factory":
  - - acseq
    - "2"
  - - cfg
    - "{\"version_control_contract\":\"contract1\",\"ans_host_contract\":\"contract0\",\"module_factory_address\":\"contract2\",\"ibc_host\":\"contract5\"}"
  - - contract_info
    - "{\"contract\":\"abstract:account-factory\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ans-host":
  - - config
    - "{\"next_unique_pool_id\":1}"
  - - contract_info
    - "{\"contract\":\"abstract:ans-host\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
  - - registered_dexes
    - "[]"
"abstract:ibc-client":
  - - config
    - "{\"version_control\":{\"address\":\"contract1\"},\"ans_host\":{\"address\":\"contract0\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-client\",\"version\":\"0.20.0\"}"
  - - module_data
    - "{\"module\":\"abstract:ibc-client\",\"version\":\"0.20.0\",\"dependencies\":[],\"metadata\":null}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:ibc-host":
  - - cfg
    - "{\"ans_host\":{\"address\":\"contract0\"},\"account_factory\":\"contract3\",\"version_control\":{\"address\":\"contract1\"}}"
  - - contract_info
    - "{\"contract\":\"abstract:ibc-host\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-0":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"¡lVë\\n»RiÂà×ÄN¾XÎMÂ\\u0004\\u0018Þng\\u000Bóq\\u001E<¤ª\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"owner\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:manager-local-1":
  - - "\u0000\u0007modulesabstract:proxy"
    - "\"\\u000F\u007fòßnHÕqa°ÊT)­M~ÉñÊY'\\u0005z\\u0019]¢þ{\\u0016\""
  - - "\u0000{12}is_suspended"
    - "false"
  - - "\u0000{4}info"
    - "{\"name\":\"Default Abstract Account\",\"governance_details\":{\"Monarchy\":{\"monarch\":\"owner\"}},\"chain_id\":\"cosmos-testnet-14002\",\"description\":null,\"link\":null}"
  - - "\u0000{6}config"
    - "{\"version_control_address\":\"contract1\",\"module_factory_address\":\"contract2\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":1}"
  - - context
    - "[]"
  - - contract_info
    - "{\"contract\":\"abstract:manager\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:module-factory":
  - - "\u0000{5}config"
    - "{\"version_control_address\":\"contract1\",\"ans_host_address\":\"contract0\"}"
  - - contract_info
    - "{\"contract\":\"abstract:module-factory\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"
"abstract:proxy-local-0":
  - - "\u0000{5}state"
    - "{\"modules\":[\"âÝ,s?&>Y\\\\\\u0014·þ 44\\u0001\\u0014Dú~;t©Þ¾1FÛ\\u0002öü\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":0}"
  - - admin
    - "\"âÝ,s?&>Y\\\\\\u0014·þ 44\\u0001\\u0014Dú~;t©Þ¾1FÛ\\u0002öü\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.20.0\"}"
"abstract:proxy-local-1":
  - - "\u0000{5}state"
    - "{\"modules\":[\"v\\u0019H%,Ð\\bð ú×i-iù^Æè\\u001F\\u0001T?¹ÆÅzãûÉ\"]}"
  - - "\u0000{6}ans_host"
    - "{\"address\":\"contract0\"}"
  - - acc_id
    - "{\"trace\":\"local\",\"seq\":1}"
  - - admin
    - "\"v\\u0019H%,Ð\\bð ú×i-iù^Æè\\u001F\\u0001T?¹ÆÅzãûÉ\""
  - - contract_info
    - "{\"contract\":\"abstract:proxy\",\"version\":\"0.20.0\"}"
"abstract:version-control":
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0005proxy0.20.0"
    - "{\"account_base\":6}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u0007manager0.20.0"
    - "{\"account_base\":5}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\bans-host0.20.0"
    - "{\"native\":\"contract0\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\nibc-client0.20.0"
    - "{\"native\":\"contract4\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000emodule-factory0.20.0"
    - "{\"native\":\"contract2\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000faccount-factory0.20.0"
    - "{\"native\":\"contract3\"}"
  - - "\u0000\u0003lib\u0000\babstract\u0000\u000fversion-control0.20.0"
    - "{\"native\":\"contract1\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0000"
    - "{\"manager\":\"âÝ,s?&>Y\\\\\\u0014·þ 44\\u0001\\u0014Dú~;t©Þ¾1FÛ\\u0002öü\",\"proxy\":\"¡lVë\\n»RiÂà×ÄN¾XÎMÂ\\u0004\\u0018Þng\\u000Bóq\\u001E<¤ª\"}"
  - - "\u0000\u0004accs\u0000\u0005local\u0000\u0000\u0000\u0001"
    - "{\"manager\":\"v\\u0019H%,Ð\\bð ú×i-iù^Æè\\u001F\\u0001T?¹ÆÅzãûÉ\",\"proxy\":\"\\u000F\u007fòßnHÕqa°ÊT)­M~ÉñÊY'\\u0005z\\u0019]¢þ{\\u0016\"}"
  - - "\u0000\u0005nmspcabstract"
    - "{\"trace\":\"local\",\"seq\":0}"
  - - "\u0000\u0007nmspc_a\u0000\u0005local\u0000\u0004\u0000\u0000\u0000\u0000abstract"
    - "8"
  - - cfg
    - "{\"account_factory_address\":\"contract3\",\"allow_direct_module_registration_and_updates\":true,\"namespace_registration_fee\":null}"
  - - contract_info
    - "{\"contract\":\"abstract:version-control\",\"version\":\"0.20.0\"}"
  - - ownership
    - "{\"owner\":\"owner\",\"pending_owner\":null,\"pending_expiry\":null}"

